feat: add a third article about the backend architecture (#364)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2022-08-25-backend-errors-connexions.md

---
layout: post

> title: "BFF's error definition, and handling connexions to multiple API"

Reviewer: ```suggestion
title: "BFF's error definition, and handling connections to multiple API"
```

Reviewer: Also the filename must be changed

description: "What is an error, a failing API? How is the BFF handling connexions to multiple API?"
author: v_claras
tags: [backend, php, api, api-gateway, back-for-front, error, timout, retry, slo, guzzle]
comments: true
color: rgb(255,128,0)
language: en
---

A ~quick~ sidetrack in [our series](#from-the-same-series) about Bedrock's API gateway.
This piece define what are we talking about when we say: an error, and explains how we handle the multiples connexions to the services we are calling.

## Definition

In [the previous article](/2022/08/12/backend-fallbacks.html), we've seen how we handle errors.
This was mainly from a business point of view, and how it's done in our domain.

*But what is "an error"?*

This term is a bit generic, and the definition will be too: *an error, is anything unexpected by the application*.

In our context of an API Gateway, we are restricting this to the services we are calling.

This can be, but not exhaustively, a service not responding because:
* it's offline;
* it's taking too much time to answer;
* it's responding with a 5** error (when talking about an API);

> * it's giving us an unexpected content.

Reviewer 2: ```suggestion
* it's giving us an invalid or unexpected content.
```

## What are the consequences of those errors?

> The first issue, is that we won't be able to display some part of the application as intended.

Reviewer 2: ```suggestion
The first issue is: we won't be able to display some part of the application as intended.
```

> We've talked about this previously already.

Reviewer 2: SI tu as un lien, avec une ancre vers le passage précis ?

> The second, more insidious error, is that it can slow down terribly our BFF.

Reviewer 2: ```suggestion
The second, more insidious, error is it can slow our BFF down terribly.
```

Pas sûr pour le ré-ordonnement de mots à la fin… Si quelqu'un d'autre a un avis ?

Author: Je pense pas. Ou alors
```suggestion
The second error, more insidious, is it can slow down our BFF terribly.
```

> The BFF response time is, in average, equals to the slowest service we are calling.

Reviewer 2: ```suggestion
The BFF response time is, on average, equal to the slowest service we are calling.
```

Reviewer 2: Ici, je note (mais je crois que ça s'appliquerait ailleurs) qu'on a deux entités :

 * "the bff"
 * "we"

Est-ce qu'en fait, on ne voudrait pas dire "the bff" / "it" dans les deux moitiés de la phrase ?

Author: Changé pour `The BFF`.

> If a service, usually responding in 200ms, starts slowing down to an average response time of 1s, while timing out half the time, it will increase the BFF response time to 1,5s (1s average, and 50% retry).

Reviewer 2: ```suggestion
If a service that usually responds in 200ms starts slowing down to an average response time of 1s and also times out half the time, it will increase the BFF response time to 1,5s (1s average, and 50% retry).
```

That's why we must be careful when configuring those timeouts.

> The BFF has SLO, and frontend applications will cut any connexion taking too long.

Reviewer 2: ```suggestion
The BFF exposes a response-time SLO, and frontend applications will cut any connection that takes too long.
```

> It's better to lose some responses, than risking slowing the BFF to a point where frontend won't get any response.

Reviewer 2: ```suggestion
Losing some parts of the responses is better than slowing the BFF down to a point where frontend won't get any response at all.
```

## How are we mitigating the errors?

> For any remote service, we make sure to use short timeout, and retry when we must.

Reviewer 2: ```suggestion
For any remote service, we configure short timeouts, and retry when we must.
```

> For any remote service, we configure short timeouts, and retry when we must.

Reviewer: what `short` means ?
1ms ?
10ms ?
100ms ?
1s ?
10s ?

-> you should give a real-world value

Author: I have asked myself the same thing while writing it :sweat_smile: , and did nothing about it ><
Here is a suggesion, I want to add the following sentence right after:
> A short timeout is a timeout as close as possible from the SLO of the called services, and that will match 99% of our calls.
Is this ok?

Reviewer: Why `as close as possible from the SLO`
Why not directly `same value as the API's SLA` ?

Author: Because sometimes, the SLA doesn't work for us. We might call a service that we know is slow, and are ok to cut half the calls. I should explicit it.

The values are tailored according to our usages.
We use our monitoring to adapt those values in order to reduce the number of errors, while minimizing the impact on the BFF response time.
We are also constantly challenging ou colleagues to improve the average response time of their services that we are calling.

The choice of using retries is based on the information criticality.
For example, retrieving the user's previous viewing sessions, is important for his/her experience, so we're using a retry here.

> On the opposite, analytics are less important, so we don't use any retry there.

Reviewer 2: Top de parler de cas où on ne fait pas de retry \o/

```yaml
    app.http_client_configs.best_effort:
        retry: 1
        timeout: 0.6
        connect_timeout: 0.1
    app.http_client_configs.fast_fail:
        retry: 0
        timeout: 0.6
        connect_timeout: 0.1
    app.http_client_configs.long_fail:
        retry: 0
        timeout: 1
        connect_timeout: 0.1
    app.http_client_configs.reliant:
        retry: 2
        timeout: 60
        connect_timeout: 0.1
```

> Above, you can see the yaml configuration used inside our Symfony application. It's used to build the Guzzle clients.

Reviewer 2: ```suggestion
Above, you can see the yaml configuration our Symfony application uses to build its Guzzle clients.
```

```yaml
    # Service definition with its aliases.
    BFF\Domain\Content\Repository: '@BFF\Domain\Content\Repository.fast_fail'
    BFF\Domain\Content\Repository.best_effort: '@BFF\Infra\HttpContentClient.best_effort'
    BFF\Domain\Content\Repository.fast_fail: '@BFF\Infra\HttpContentClient.fast_fail'

    # Concrete implementations
    BFF\Infra\HttpContentClient.best_effort:
        class: 'BFF\Infra\HttpContentClient'
        factory: ['@BFF\Infra\ContentClientFactory', 'create']
        bind:
            $clientConfig: '%app.http_client_configs.best_effort%'
    BFF\Infra\HttpContentClient.fast_fail:
        class: 'BFF\Infra\HttpContentClient'
        factory: ['@BFF\Infra\ContentClientFactory', 'create']
        bind:
            $clientConfig: '%app.http_client_configs.fast_fail%'

    # Other services using the Repository
    BFF\Domain\Navigation\NavBarResolver:
        $content: '@BFF\Domain\Content\Repository.best_effort'

    BFF\Domain\Layout\BlockResolver:
        $content: '@BFF\Domain\Content\Repository.fast_fail'
```

_This is an over simplified example as we have more layers and wrappers used for things like caching, monitoring, logging, etc._

## Conclusion

> In this article, we've clarified what an error is, and explained that we cannot generalize the configuration and usage of our APIs.

Reviewer 2: ```suggestion
In this article, we've clarified what an error is, and explained that we cannot generalize the configuration and usage of our APIs. Timeouts and retries, especially, must be tailored depending on the criticality of each call.
```

Hu, marrant, Antidote connait le verbe "to retry", mais n'a pas l'air de connaitre "retry" (et "retries") comme nom :-/

> This was a deviation on the road to our next article, where we will talk about monitoring the errors and stopping calls to failing APIs.

Reviewer 2: "stopping calls", j'imagine via un circuit breaker? Mettre le terme pourrait faire envie, en mode teasing?

Author: Mais à la fois ça spoil non ? C'est là où j'ai un doute :p

Reviewer: quelqu'un qui a lu cet article en entier, l'a fait pour savoir "comment" tu as fait.
Ca sera pareil pour les circuit-breakers IMHO

Author: J'ai pas compris ta réponse :sweat_smile:

Reviewer: Les gens qui prennent le temps de lire l'article en entier, comprennent le pourquoi, mais sont aussi intéressés par comprendre le comment tu as abouti à ce résultat.
Parler de circuit-breaker, ça va leur parler et leur donner envie de lire l'article, car ils vont vouloir comprendre "comment" tu as mis en place un circuit-breaker, c'est ça la plus-value. Savoir que tu as mis un circuit breaker, c'est comme mettre un timeout. On s'en fou, ça n'apporte pas de valeur. Dire la valeur du timeout et pourquoi, c'est ça qui est intéressant IMHO.
